File: tests/syquex/test_mapa.py
"""
FASE 24 — Test de Mapa hash (Manual 3 §5.2).

TDD: este test ES la especificación. Si _syn_mapa_* no existen,
el C test NO compila — eso es correcto. Se corrige el CÓDIGO, no el test.

Manual 3 §5.2: Mapa<K,V> — diccionario hash
Comando: pytest tests/syquex/test_mapa.py -v
Criterio: 0 fugas, operaciones correctas
"""
import os
import subprocess
import sys

import pytest

pytestmark = pytest.mark.syquex

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))

BIN_NAME = "test_mapa.exe"
BIN_ABS = os.path.join(PROJECT_ROOT, "tests", BIN_NAME)
TEST_SRC = os.path.join(PROJECT_ROOT, "tests", "test_mapa.c")


def _find_gcc():
    gcc_candidate = os.path.join(PROJECT_ROOT, "toolchain_gcc12", "mingw64", "bin", "gcc.exe")
    if os.path.exists(gcc_candidate):
        return gcc_candidate
    for candidate in ("gcc", "gcc.exe"):
        try:
            subprocess.run([candidate, "--version"], capture_output=True)
            return candidate
        except FileNotFoundError:
            continue
    return gcc_candidate


@pytest.fixture(scope="module")
def exe_path():
    """Compila el test C y retorna la ruta al ejecutable."""
    if os.path.exists(BIN_ABS):
        return BIN_ABS

    gcc = _find_gcc()
    r = subprocess.run(
        [gcc, "-O2", "-I", PROJECT_ROOT, "-I.", "-o", BIN_ABS, TEST_SRC,
         os.path.join(PROJECT_ROOT, "synapse_rt_memory.o"),
         "-lm", "-lpthread", "-lws2_32"],
        capture_output=True, text=True, timeout=120
    )
    if r.returncode != 0:
        pytest.fail(
            f"test_mapa.c NO COMPILA (TDD: falta implementar §5.2).\n"
            f"Error de gcc:\n{r.stderr[-1000:]}"
        )
    assert os.path.exists(BIN_ABS), f"{BIN_NAME} no se creó"
    return BIN_ABS


class TestMapa:
    """Manual 3 §5.2 — Mapa hash: 0 fugas, operaciones correctas."""

    def test_creacion(self, exe_path):
        """Mapa se crea vacío."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert r.returncode == 0, f"test_mapa falló:\n{r.stdout}\n{r.stderr}"
        assert "mapa_crear retorna no-NULL" in r.stdout
        assert "mapa vacío tiene longitud 0" in r.stdout

    def test_poner_obtener(self, exe_path):
        """Poner y obtener pares clave-valor."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "longitud = 3 tras 3 poner" in r.stdout
        assert 'obtener("nombre") == 42' in r.stdout
        assert 'obtener("edad") == 28' in r.stdout
        assert 'obtener("activo") == 1' in r.stdout

    def test_contiene(self, exe_path):
        """Verificar existencia de claves."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert 'contiene("nombre") == true' in r.stdout
        assert 'contiene("inexistente") == false' in r.stdout

    def test_actualizar(self, exe_path):
        """Actualizar valor de clave existente."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert 'actualizar "nombre" a 99' in r.stdout
        assert "longitud sigue en 3 tras actualizar" in r.stdout

    def test_eliminar(self, exe_path):
        """Eliminar par clave-valor."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert 'longitud = 2 tras eliminar "edad"' in r.stdout
        assert 'contiene("edad") == false tras eliminar' in r.stdout

    def test_limpiar(self, exe_path):
        """Limpiar todo el mapa."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "longitud = 0 tras limpiar" in r.stdout

    def test_expansion(self, exe_path):
        """500 claves — expansión automática."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "longitud = 500" in r.stdout
        assert "key_0 == 0" in r.stdout
        assert "key_250 == 250" in r.stdout
        assert "key_499 == 499" in r.stdout

    def test_claves_valores(self, exe_path):
        """Obtener listas de claves y valores."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "claves() retorna no-NULL" in r.stdout
        assert "valores() retorna no-NULL" in r.stdout
        assert "claves tiene 3 elementos" in r.stdout
        assert "valores tiene 3 elementos" in r.stdout

    def test_null_safety(self, exe_path):
        """NULL safety — no crashea con NULL."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "longitud(NULL) == 0" in r.stdout
        assert "poner(NULL) no crashea" in r.stdout

    def test_no_leaks(self, exe_path):
        """0 fugas."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "0 failed" in r.stdout
        assert "RESULTADO" in r.stdout

File: tests/test_mapa.c
// FASE 24 — Test de Mapa hash (Manual 3 §5.2)
// TDD: este test ES la especificación. Si las funciones _syn_mapa_*
// no existen, el test NO compila — eso es correcto.
//
// Manual 3 §5.2: Mapa<K,V> — diccionario hash
// Comando: pytest tests/syquex/test_mapa.py -v
// Criterio: 0 fugas, operaciones correctas

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "synapse_rt_types.h"

static int passed = 0;
static int failed = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

int main(void) {
    setbuf(stdout, NULL);

    // === 1. Creación y básicos ===
    printf("=== 1. Creacion y operaciones basicas ===\n");
    void* m = _syn_mapa_crear();
    CHECK(m != NULL, "mapa_crear retorna no-NULL");
    CHECK(_syn_mapa_longitud(m) == 0, "mapa vacío tiene longitud 0");

    // === 2. Poner y obtener ===
    printf("=== 2. Poner y obtener ===\n");
    _syn_mapa_poner(m, "nombre", 42);
    _syn_mapa_poner(m, "edad", 28);
    _syn_mapa_poner(m, "activo", 1);
    CHECK(_syn_mapa_longitud(m) == 3, "longitud = 3 tras 3 poner");

    CHECK(_syn_mapa_obtener(m, "nombre") == 42, "obtener(\"nombre\") == 42");
    CHECK(_syn_mapa_obtener(m, "edad") == 28, "obtener(\"edad\") == 28");
    CHECK(_syn_mapa_obtener(m, "activo") == 1, "obtener(\"activo\") == 1");

    // === 3. Contiene ===
    printf("=== 3. Contiene ===\n");
    CHECK(_syn_mapa_contiene(m, "nombre") == 1, "contiene(\"nombre\") == true");
    CHECK(_syn_mapa_contiene(m, "inexistente") == 0, "contiene(\"inexistente\") == false");

    // === 4. Actualizar valor ===
    printf("=== 4. Actualizar valor ===\n");
    _syn_mapa_poner(m, "nombre", 99);
    CHECK(_syn_mapa_obtener(m, "nombre") == 99, "actualizar \"nombre\" a 99");
    CHECK(_syn_mapa_longitud(m) == 3, "longitud sigue en 3 tras actualizar");

    // === 5. Eliminar ===
    printf("=== 5. Eliminar ===\n");
    _syn_mapa_eliminar(m, "edad");
    CHECK(_syn_mapa_longitud(m) == 2, "longitud = 2 tras eliminar \"edad\"");
    CHECK(_syn_mapa_contiene(m, "edad") == 0, "contiene(\"edad\") == false tras eliminar");
    CHECK(_syn_mapa_obtener(m, "edad") == 0, "obtener(\"edad\") == 0 tras eliminar");

    // === 6. Limpiar ===
    printf("=== 6. Limpiar ===\n");
    _syn_mapa_limpiar(m);
    CHECK(_syn_mapa_longitud(m) == 0, "longitud = 0 tras limpiar");
    CHECK(_syn_mapa_contiene(m, "nombre") == 0, "contiene(\"nombre\") == false tras limpiar");

    // === 7. Muchas claves (expansión) ===
    printf("=== 7. Expansion (500 claves) ===\n");
    char clave[32];
    for (int i = 0; i < 500; i++) {
        snprintf(clave, sizeof(clave), "key_%d", i);
        _syn_mapa_poner(m, clave, i);
    }
    CHECK(_syn_mapa_longitud(m) == 500, "longitud = 500");
    CHECK(_syn_mapa_obtener(m, "key_0") == 0, "key_0 == 0");
    CHECK(_syn_mapa_obtener(m, "key_250") == 250, "key_250 == 250");
    CHECK(_syn_mapa_obtener(m, "key_499") == 499, "key_499 == 499");
    CHECK(_syn_mapa_contiene(m, "key_100") == 1, "contiene key_100");

    // === 8. Claves y valores ===
    printf("=== 8. Claves y valores ===\n");
    _syn_mapa_limpiar(m);
    _syn_mapa_poner(m, "a", 1);
    _syn_mapa_poner(m, "b", 2);
    _syn_mapa_poner(m, "c", 3);
    void* claves = _syn_mapa_claves(m);
    void* valores = _syn_mapa_valores(m);
    CHECK(claves != NULL, "claves() retorna no-NULL");
    CHECK(valores != NULL, "valores() retorna no-NULL");
    CHECK(_syn_lista_longitud(claves) == 3, "claves tiene 3 elementos");
    CHECK(_syn_lista_longitud(valores) == 3, "valores tiene 3 elementos");

    // Verificar que los valores están (el orden puede variar por hash)
    int64_t v0 = _syn_lista_obtener(valores, 0);
    int64_t v1 = _syn_lista_obtener(valores, 1);
    int64_t v2 = _syn_lista_obtener(valores, 2);
    CHECK((v0 == 1 || v0 == 2 || v0 == 3), "valor[0] es 1, 2 o 3");
    CHECK((v1 == 1 || v1 == 2 || v1 == 3), "valor[1] es 1, 2 o 3");
    CHECK((v2 == 1 || v2 == 2 || v2 == 3), "valor[2] es 1, 2 o 3");

    _syn_lista_liberar(claves);
    _syn_lista_liberar(valores);

    // === 9. NULL safety ===
    printf("=== 9. NULL safety ===\n");
    CHECK(_syn_mapa_longitud(NULL) == 0, "longitud(NULL) == 0");
    _syn_mapa_poner(NULL, "x", 1);  // no crashea
    CHECK(1, "poner(NULL) no crashea");
    CHECK(_syn_mapa_obtener(NULL, "x") == 0, "obtener(NULL, \"x\") == 0");
    CHECK(_syn_mapa_contiene(NULL, "x") == 0, "contiene(NULL, \"x\") == 0");

    // === 10. Liberación ===
    printf("=== 10. Liberacion ===\n");
    _syn_mapa_liberar(m);
    CHECK(1, "mapa_liberar no crashea");

    printf("\n=== RESULTADO: %d passed, %d failed ===\n", passed, failed);
    return failed > 0 ? 1 : 0;
}
